src/test/java/bsh/Namespace_Chaining_Test.java

```diff
@@ -48,6 +48,45 @@ public void namespace_nesting() throws UtilEvalError {
         assertEquals(42, root.getVariable("bar"));
     }
 
+   @Test
+   public void override_child_namespace() throws Exception {
+        Interpreter root = new Interpreter();
+        Interpreter child = new Interpreter(new StringReader(""), System.out, System.err, false, new NameSpace(root.getNameSpace(), "child"));
+
+        root.eval("int bar=42;");
+        child.eval("int bar=4711;");
+
+        assertEquals(42, root.eval("bar;"));
+        assertEquals(4711, child.eval("bar;"));
+
+        // Method foo declared in parent namespace returns bar
+        root.eval("int foo() { return bar; }");
+
+        assertEquals(42, root.eval("foo();"));
+        assertEquals(42, child.eval("foo();"));
+    }
+
+    @Test
+    public void override_child_child_namespace() throws Exception {
+        Interpreter root = new Interpreter();
+        Interpreter child = new Interpreter(new StringReader(""), System.out, System.err, false, new NameSpace(root.getNameSpace(), "child"));
+        Interpreter child2 = new Interpreter(new StringReader(""), System.out, System.err, false, new NameSpace(child.getNameSpace(), "child2"));
+
+        root.eval("int bar=42;");
+        child.eval("int bar=4711;");
+        child2.eval("int bar=5;");
+
+        assertEquals(42, root.eval("bar;"));
+        assertEquals(4711, child.eval("bar;"));
+        assertEquals(5, child2.eval("bar;"));
+
+        // Method foo declared in parent namespace returns bar
+        root.eval("int foo() { return bar; }");
+
+        assertEquals(42, root.eval("foo();"));
+        assertEquals(42, child.eval("foo();"));
+        assertEquals(42, child2.eval("foo();"));
```

Comment:
But this is not correct, the method is declared in parent namespace, correct but it is returning the variable `bar`.

Which for `root` is 42, but for `child` has the value of 4711 and for `child2` should be 5.

There is inheritance but gets overwritten if declared or changed.

Lets look at it from another perspective:

```java
class Root {
   int bar = 42;
   int foo() {
      return bar;
   }
}

class Child extends Root {
   int bar = 4711;
}

int baz = new Child().foo();
```

What will the value be for `baz`?

Reply:
The value will be 4711.  

This bug has nothing to do with inheritance, and this reversion will not change your example from working.  The change that this will revert was looking up and using variables in the namespace where the scripted method was being executed in, rather than using the variables from the namespace that the function was declared in.   This behaviour causes the namespace stack to become irrecoverably corrupted and subsequently run scripts to fail so it is a serious bug.  

This bug has been open for almost 4 years, it is a showstopper, and it is blocking other fixes from being merged.   Time to revert and close off this issue.

Additionally I would say that the behaviour was a bit weird anyway.  If there is interest in making a fix to bring this back I think that:
1. there should be a discussion about the proposed behavioural change.
2. all the unit tests should pass.